### PR TITLE
[SPARK-42093][SQL] Move JavaTypeInference to AgnosticEncoders

### PR DESCRIPTION
**Pull request by hvanhovell**
### What changes were proposed in this pull request?
This PR makes `JavaTypeInference` produce an `AgnosticEncoder`. The expression generation for these encoders is moved to `ScalaReflection`.

### Why are the changes needed?
For the Spark Connect Scala Client we also want to be able to use Java Bean based results.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
I have added a lot of tests to `JavaTypeInferenceSuite`.

### CODE REVIEW

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Could you check Scala 2.13 too? The original `AgnosticEncoders` PR broke Scala 2.13 with Kafka SQL module.
- https://github.com/apache/spark/pull/39517#pullrequestreview-1250710992

**Comment by cloud-fan** [comment]
yea let's fix scala 2.13 first

**Comment by hvanhovell** [comment]
@cloud-fan @dongjoon-hyun can you take another look please?

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master/3.4!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala [hunk 1]
```diff
@@ -166,317 +148,58 @@ object JavaTypeInference {
...
-              nullable = true,
-              serializerFor(_, valueType))
-          )
-
-        case other if other.isEnum =>
-          createSerializerForString(
-            Invoke(inputObject, "name", ObjectType(classOf[String]), returnNullable = false))
-
-        case other =>
-          val properties = getJavaBeanReadableAndWritableProperties(other)
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala:465]: Hmm, I'm not sure if this has been discussed, but this is a breaking change causing customer issue like https://issues.apache.org/jira/browse/SPARK-48073.

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala:465]: Is this an intentional change? @hvanhovell

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala:465]: No matter what, this change has been there for a while. It seems difficult to change it back to old behavior. As discussed in the JIRA, I'd like to see if we can add a SQL conf to fall back to previous behavior of `Encoder.bean()`.

Comment (hvanhovell) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala:465]: The behavior was apparently there before I got to it, this is PR that fixed it later: https://github.com/apache/spark/pull/42829

For the record there is no specification here, only tests.

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala:465]: Thanks. After more internal discussion with the customer, we decided not to make a change to previous behavior with a config.

**Pull request closed without merging**